test/cpp/util/spiffe_test_credentials.h

```diff
@@ -0,0 +1,128 @@
+/*
+ *
+ * Copyright 2019 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#ifndef GRPC_TEST_CPP_UTIL_SPIFFE_TEST_CREDENTIALS_H
+#define GRPC_TEST_CPP_UTIL_SPIFFE_TEST_CREDENTIALS_H
+
+#include <grpcpp/security/credentials.h>
+#include <grpcpp/security/server_credentials.h>
+#include <memory>
+#include <thread>
+#include <vector>
+#include "test/cpp/util/test_credentials_provider.h"
+
+namespace grpc {
+namespace testing {
+
+const char kSpiffeCredentialsType[] = "spiffe";
+
+/** This class manages all the threads spawned by a
+ *  SpiffeCredentialTypeProvider with asynchronous server authorization check
+ *  enabled. **/
+class SpiffeThreadList {
+ public:
+  ~SpiffeThreadList() {
+    for (size_t i = 0; i < thread_list_.size(); i++) {
+      thread_list_[i].join();
+    }
+  }
+
+  void add_thread(std::thread t) { thread_list_.push_back(std::move(t)); }
+
+ private:
+  std::vector<std::thread> thread_list_;
+};
+
+/** This method creates a TlsCredentialsOptions instance with no key materials,
+ *  whose credential reload config is configured using the
+ *  TestSyncTlsCredentialReload class, and whose server authorization check
+ *  config is determined as follows:
+ *  - if |is_client| is true,
+ *      -if |is_async|, configured by TestAsyncTlsServerAuthorizationCheck,
+ *      -otherwise, configured by TestSyncTlsServerAuthorizationCheck.
+ *  - otherwise, the server authorization check config is not populated.
+ *
+ *  Further, the cert request type of the options instance is always set to
+ *  GRPC_SSL_REQUEST_AND_REQUIRE_CLIENT_CERTIFICATE_AND_VERIFY for both the
+ *  client and the server. **/
```

Comment:
I think `GRPC_SSL_REQUEST_AND_REQUIRE_CLIENT_CERTIFICATE_AND_VERIFY` will only take effect at server side, not client side. Since client will always validate server's certificate.

Reply:
See my comment above.